simulator/recorder/recorder.go

```diff
@@ -0,0 +1,179 @@
+package recorder
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"time"
+
+	"golang.org/x/xerrors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+type Event string
+
+var (
+	Add    Event = "Add"
+	Update Event = "Update"
+	Delete Event = "Delete"
+)
+
+const defaultRecordBatchCapacity = 1000
+
+type Service struct {
+	client              dynamic.Interface
+	gvrs                []schema.GroupVersionResource
+	path                string
+	recordCh            chan Record
+	recordBatchCapacity int
+}
+
+type Record struct {
+	Time     time.Time                 `json:"time"`
+	Event    Event                     `json:"event"`
+	Resource unstructured.Unstructured `json:"resource"`
+}
+
+var DefaultGVRs = []schema.GroupVersionResource{
+	{Group: "", Version: "v1", Resource: "namespaces"},
+	{Group: "scheduling.k8s.io", Version: "v1", Resource: "priorityclasses"},
+	{Group: "storage.k8s.io", Version: "v1", Resource: "storageclasses"},
+	{Group: "", Version: "v1", Resource: "persistentvolumeclaims"},
+	{Group: "", Version: "v1", Resource: "nodes"},
+	{Group: "", Version: "v1", Resource: "persistentvolumes"},
+	{Group: "", Version: "v1", Resource: "pods"},
+}
+
+type Options struct {
+	GVRs                []schema.GroupVersionResource
+	RecordFile          string
+	RecordBatchCapacity *int
+}
+
+func New(client dynamic.Interface, options Options) *Service {
+	gvrs := DefaultGVRs
+	if options.GVRs != nil {
+		gvrs = options.GVRs
+	}
+
+	recordBatchCapacity := defaultRecordBatchCapacity
```

Comment:
Hmm. Having the capacity on `recordCh` doesn't make much sense because if it's blocked, the event handler would have the pending events (iiuc). So, it's just shifting the pending events from the event handler to `recordCh`.
I'd rather implement the bulk insert into the file.

Reply:
Implemented bulk insert

Reply:
It still doesn't smell quite right because we'd block the write to the channel when it hits the cap and it's writing to the file.

I guess the channel isn't needed in the first place, it should be like:
1. Event handler stores events into `records` slice. (acquire lock)
2. The goroutine reads `records` slice _periodically_ (acquire rlock), and writes _all_ events into the file (note: when writing to the file, it already released the rlock).

Then, we can have a configuration param to for the interval of (2), not the capacity. (the default interval would be 5s or something.)

Reply:
Fixed to use slice for buffering.